### ISSUE TITLE
[HOLD for payment 2023-05-23] [$2000] Android - Sign in - The keyboard opens and disappears for a brief moment

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#16357]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Issue found when executing PR** https://github.com/Expensify/App/pull/15731

## Action Performed:
. Open the App
2. Navigate to the login page
3. Press the Create a new account link in the footer
4. Verify that you are navigated to the login page
5. Enter an email address and press Continue
6. Scroll down to the footer and press Log in
7. Verify that you are navigated back to the login page

## Expected Result:
The keyboard doesn't open

## Actual Result:
The keyboard opens and disappears for a brief moment

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.2.88.0

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/226730725-c1544b8c-7904-4011-80c0-4cddab18f712.mp4

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01f770760fee7155f3</li>
        <li>Upwork Job ID: 1640402122644557824</li>
        <li>Last Price Increase: 2023-04-10</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @tjferriss (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01f770760fee7155f3

**Comment by tjferriss** [comment]
Reproduced via BrowserStack

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @Santhosh-Sellavel (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @roryabraham (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Keyboard is shown for a while when we go to login page from footer link while we are at password/2fa form.

### What is the root cause of that problem?

Currently, when we go to the password form, the login form is not unmounted, but just hidden.

https://github.com/Expensify/App/blob/10d3d52e993303db477478d323af925f4d439acf/src/pages/signin/SignInPage.js#L111-L113

The login form will only be shown if this condition meets

https://github.com/Expensify/App/blob/10d3d52e993303db477478d323af925f4d439acf/src/pages/signin/SignInPage.js#L56-L59

When we press the Log in footer link, it will call this `navigateHome`. It does 2 thing, `navigate` and `clearSignInData`,
https://github.com/Expensify/App/blob/10d3d52e993303db477478d323af925f4d439acf/src/pages/signin/SignInPageLayout/Footer.js#L24-L29

Even though we navigate to the same page, the page will remount. However, if we only do the navigate, the password form will still be shown. That's why we need the `clearSignInData`.

Now, inside `LoginForm`, we auto focus the input every time the form becomes visible.

https://github.com/Expensify/App/blob/10d3d52e993303db477478d323af925f4d439acf/src/pages/signin/LoginForm.js#L93-L96

So, while we are navigating with `navigateHome`, the `LoginForm` will be updated with the new value of `isVisible` is true (because of `clearSignInData`) and call the focus, thus the keyboard shows. After the navigation is done, the focus is lost. This makes the keyboard shows for a moment.

### What changes do you think we should make in order to solve the problem?


The solution is to prevent the focus to happen as we are leaving the page, going to the same page. To prevent it, we can check if the current screen/page is focused or not.

1. Wrap `LoginForm` with `withNavigationFocus`
2. Check inside the `componentDidUpdate`, if the screen is not focused anymore, return early.

https://github.com/Expensify/App/blob/10d3d52e993303db477478d323af925f4d439acf/src/pages/signin/LoginForm.js#L93-L96

**Comment by Santhosh-Sellavel** [comment]
cc: @luacmartins do we need to treat this as regression?

**Comment by Santhosh-Sellavel** [comment]
@bernhardoj I'm not sure why this happens on android alone root cause doesn't explain this clearly

**Comment by bernhardoj** [comment]
Adding explanation to it on my [proposal](https://github.com/Expensify/App/issues/16357#issuecomment-1485537416).

**Comment by Santhosh-Sellavel** [comment]
Sorry for the delay. @roryabraham Your thoughts on the [proposal](https://github.com/Expensify/App/issues/16357#issuecomment-1485537416)?

**Comment by hellohublot** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Android - Sign in - The keyboard opens and disappears for a brief moment

### What is the root cause of that problem?

https://github.com/Expensify/App/blob/4ee21e95a38858df28e46caf54dc97e1bca9bc7a/src/pages/signin/SignInPageLayout/Footer.js#L24-L29


Why does the keyboard suddenly appear ?
- `Session.clearSignInData()` will cause the old page `SignInPage.LoginForm.isVisible = true`, so the old page `SignInPage.LoginForm.TextInput` will be focus in `componentDidUpdate`

---

Why did the keyboard disappear so quickly again ?
- Because `Navigation.navigate` will create a `new SignInPage page`, and also regenerate a `SignInPage.LoginForm.TextInput`, the keyboard of the `old page` will be `destroyed` accordingly
---


### What changes do you think we should make in order to solve the problem?

If we `want the keyboard to appear`, just like clicking `Go Back`, then we don’t create a new page. whether it is `Go Back` or `Create A New Account`, we need to scroll to the top
https://github.com/Expensify/App/blob/4ee21e95a38858df28e46caf54dc97e1bca9bc7a/src/pages/signin/SignInPageLayout/Footer.js#L24-L29
``` diff
-   Navigation.navigate();
```
https://github.com/Expensify/App/blob/4ee21e95a38858df28e46caf54dc97e1bca9bc7a/src/pages/signin/SignInPageLayout/index.js#L28-L30
``` diff
+   const scrollViewRef = useRef()
+   useEffect(() => {
+       scrollViewRef. current. scrollTo({ y: 0 })
+   }, [props. welcomeText, props. shouldShowWelcomeText])
```

### What alternative solutions did you explore? (Optional)
If we `don’t want the keyboard to appear`, `Go Back` and `Create A New Account` need to be handled differently here, but the data used for judgment is still the same, so we may need a new condition, which will increase the logic

**Comment by MelvinBot** [comment]
@tjferriss, @roryabraham, @Santhosh-Sellavel Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by Santhosh-Sellavel** [comment]
> Sorry for the delay. @roryabraham Your thoughts on the [proposal](https://github.com/Expensify/App/issues/16357#issuecomment-1485537416)?

@roryabraham bump!

**Comment by hellohublot** [comment]
@Santhosh-Sellavel 

Hello, Thank you for your suggestion, but I actually also want to solve `go back` problems. 

- On the Native platform, clicking `Go Back` will only focus but not automatically scroll like the Web platform
- The same goes for the `Back` button.
- In addition, if we click `Forgot`, since there is no `input.focus`, it will not automatically scroll even on the web platform

https://user-images.githubusercontent.com/20135674/229667925-379c515b-d298-426e-8780-2475fc54a813.mp4

**Comment by MelvinBot** [comment]
@tjferriss @roryabraham @Santhosh-Sellavel this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by Santhosh-Sellavel** [comment]
@hellohublot 
> On the Native platform, clicking Go Back will only focus but not automatically scroll like the Web platform

That is a different issue, we have set the focus on the input but it is not visible on native iOS.

> In addition, if we click Forgot, since there is no input.focus, it will not automatically scroll even on the web platform

This is not a bug because it the user was intentionally scrolling-up and after performing an action here has nothing focusable. It would be kind of over-engineering i.e to handle this by manually scrolling up.

**Comment by hellohublot** [comment]
@Santhosh-Sellavel  Thanks for your suggestion, I added two solutions B and C, it seems that solution B still needs to call the `ScrollView.scrollToTop`

**Comment by MelvinBot** [comment]
@tjferriss, @roryabraham, @Santhosh-Sellavel Eep! 4 days overdue now. Issues have feelings too...

**Comment by MelvinBot** [comment]
@tjferriss, @roryabraham, @Santhosh-Sellavel Huh... This is 4 days overdue. Who can take care of this?

**Comment by Santhosh-Sellavel** [comment]
@roryabraham 

I'm not feeling strongly toward any proposals, so far Can you post your thoughts here?

We have two proposals.
[One](https://github.com/Expensify/App/issues/16357#issuecomment-1485537416) from @bernhardoj and [other](https://github.com/Expensify/App/issues/16357#issuecomment-1493072701) from @hellohublot

**Comment by bernhardoj** [comment]
@Santhosh-Sellavel there is a PR https://github.com/Expensify/App/pull/17119 which will scroll to the top instead of navigating which contradicts this statement https://github.com/Expensify/App/issues/16357#issuecomment-1494945055.

**Comment by MelvinBot** [comment]
@tjferriss @roryabraham @Santhosh-Sellavel this issue is now 3 weeks old. There is one more week left before this issue <a href='https://stackoverflowteams.com/c/expensify/questions/15233/15234#15234'>breaks WAQ</a> and will need to go internal. What needs to happen to get a PR in review this week? Please create a thread in #expensify-open-source to discuss. Thanks!

**Comment by Santhosh-Sellavel** [comment]
@roryabraham This waits for you,

We have a [proposal](https://github.com/Expensify/App/issues/16357#issuecomment-1485537416) from @bernhardoj and were confused with the root cause and solution initially. But I was okay with their proposal after adding [explaination](https://github.com/Expensify/App/issues/16357#issuecomment-1486128756) and was waiting for your input from [here](https://github.com/Expensify/App/issues/16357#issuecomment-1490705635). 

Meanwhile, we also have another proposal from @hellohublot [here](https://github.com/Expensify/App/issues/16357#issuecomment-1493072701)

IMO manually scrolling up here is kind of unnecessary, as this issue was specifically on Android and it could lead to unintended behavior on other platforms.

Now a similar solution has been used here on #17119 to address a different issue.  I'm kind of confused here, and could use your help here to move this forward, Let us know your thoughts thanks!

**Comment by roryabraham** [comment]
Ok, interesting. I too was skeptical about the scroll-to-0 solution, but after seeing the solution in action I think it actually makes a lot of sense. Whenever we're clicking on any of these buttons (`go back`, `forgot`, etc...) the user almost always wants to be scrolled to the top of the screen and have their keyboard opened. So I think I'm going to award this one to @hellohublot

**Comment by MelvinBot** [comment]
📣 @hellohublot You have been assigned to this job by @roryabraham!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by hellohublot** [comment]
@roryabraham  @Santhosh-Sellavel 

Hello. Thank you very much for your time

After I removed `Footer.navigateHome.Navigation.navigate()`, I found another problem. When I click `Create A New Account`, the console will report an error `Hoverable.onBlur.this.wrapperView.contains -> this.wrapperView is null`

Because `username.input` will be `focused`, so `Footer.Hoverable.onBlur` will be `responded`, but at this time, the `Hoverable` is also `refreshing`, so `Hoverable.View.wrapperView.ref` will become `empty` for a short time

So when you click each `Hoverable` to refresh the `Hoverable` and `focus on other inputs`, this problem will occur. 

I feel that this is a problem that the `Hoverable` does not judge the wrapperView `emptyness`. I don’t know whether I should submit a PR first to judge whether the `wrapperView is empty`.


---

Finally, here is a [PR](https://github.com/Expensify/App/pull/17119), It also has the problem above, which conflicts with my proposal, looks like my proposal can solve his problem too, and I need your help to negotiate whether to go with his proposal or mine, Thanks

**Comment by Santhosh-Sellavel** [comment]
@hellohublot Seems that's merged now can spun a PR handle the missed out changes here, thanks!

**Comment by hellohublot** [comment]
Thanks, PR is ready, https://github.com/Expensify/App/pull/17446 and https://github.com/Expensify/App/pull/17445

**Comment by Santhosh-Sellavel** [comment]
@hellohublot I'm not sure why there are two PRs, also it's says that address
> This PR is to fix https://github.com/Expensify/App/pull/17119 and https://github.com/Expensify/App/pull/17445 will print errors in the console,

Where #17445 which is an active PR?

**Comment by hellohublot** [comment]
@Santhosh-Sellavel 

Thanks for the reminder, I thought it was two different things, so it would be clearer and more manageable to separate them, if it is really not necessary, I will delete the other PR, and then merge them into the same PR

**Comment by Santhosh-Sellavel** [comment]
Please merge them

**Comment by hellohublot** [comment]
@Santhosh-Sellavel 
Thanks, I've closed another one and merged into this [one](https://github.com/Expensify/App/pull/17445)

**Comment by MelvinBot** [comment]
@tjferriss, @hellohublot, @roryabraham, @Santhosh-Sellavel Huh... This is 4 days overdue. Who can take care of this?

**Comment by Santhosh-Sellavel** [comment]
On hold now for other issue, check PR for more.

**Comment by tjferriss** [comment]
On hold for this PR https://github.com/Expensify/App/pull/17812

**Comment by MelvinBot** [comment]
@tjferriss, @hellohublot, @roryabraham, @Santhosh-Sellavel Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by Santhosh-Sellavel** [comment]
We are back on track here, and the review is in progress.

**Comment by tjferriss** [comment]
Are we still reviewing the PR?

**Comment by hellohublot** [comment]
Yes, just waiting for Expensify Reviewer to merge https://github.com/Expensify/App/pull/17445

**Comment by roryabraham** [comment]
Merged!

**Comment by tjferriss** [comment]
@Santhosh-Sellavel @hellohublot can you please both apply to the job here: https://www.upwork.com/jobs/~01060f20aaf9a83d49? 

There is a 50% penalty for this issue:
<img width="739" alt="Screenshot 2023-05-17 at 17 34 32" src="https://github.com/Expensify/App/assets/8292562/7470cc50-2d6e-4c81-aa71-ff9bb7e5348a">

**Comment by roryabraham** [comment]
Looking at the history here:

- @Santhosh-Sellavel [asked to put the PR on hold](https://github.com/Expensify/App/pull/17445#issuecomment-1518241020) April 21st
- The [PR it was on HOLD for](https://github.com/Expensify/App/pull/17812) was eventually just closed on May 3rd.

So that's 8 business days when the PR was on HOLD. Then @Santhosh-Sellavel [final-approved the PR](https://github.com/Expensify/App/pull/17445#pullrequestreview-1415947372) on May 7th, and I merged it on May 11th, so that's an additional 4 business days it was delayed outside of the contributors' control here.

So given that there were 22 business days elapsed between assignment and merge, and 12 of those days were caused by factors outside the contributors' control. That still leaves **10 business days between assignment and merge _(after correction)_**.

According to our [contributing guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#payment-for-contributions):

> Merged PR within 9 business days - 50% penalty

So by my reckoning the 50% penalty should still apply here.

**Comment by Santhosh-Sellavel** [comment]
@roryabraham 

I agree it took around 10 days. I am requesting you to waive off here as this was on hold. The hold is the problem where I lost rack of time. Post Off hold I spent 4 days reviewing and testing the PR. That's the average review time (3 - 4 days) for any PR. I had to ensure this does not cause any regressions. Thanks!

cc: @tjferriss

**Comment by hellohublot** [comment]
@tjferriss 
Thanks, I already submitted a proposal on Upwork, but I guess you can't find it (I've encountered it before), so you can send me an offer directly on payment day, thanks again


@roryabraham 
Thank you for your feedback. After May 3 and May 4 Hold status was cancelled, since the main code has changed a lot, I and the C+ reviewer need to modify all the test steps again, take screenshots and test all the modifications again, the workload of these modifications is already equivalent to another complete PR

In fact, this is the first time I got -50% penalty, but every time Expensify reviewer and C+ reviewer ask questions, every time I reply all messages within 1 day, yes I always push progress within 1 day , and this problem also consumed me a lot of time and energy, but still got -50% penalty

**Comment by roryabraham** [comment]
> since the main code has changed a lot, I and the C+ reviewer need to modify all the test steps again

I think this is a fair point – the delay added additional work due to merge conflicts. Given that you only missed the mark by 1 day, it's reasonable to assume that you would've completed it in time without the additional work caused by the delay. You were also responsive throughout, and demonstrated a good amount of urgency to push the PR forward.

So given the specific context here, I'm willing to make an exception and void the 50% penalty because it was more my fault than yours.

So, @tjferriss if you agree, that means that we should pay the full bounty amount here.

**Comment by tjferriss** [comment]
Sorry for the delay as I've been out of office.

I agree with paying the full bounty. Thanks for handling the discussion @roryabraham. 

>Thanks, I already submitted a proposal on Upwork, but I guess you can't find it (I've encountered it before), so you can send me an offer directly on payment day, thanks again

@hellohublot I don't fully follow your comment. Are you not able to apply to the job posting here: https://www.upwork.com/jobs/~01060f20aaf9a83d49?

Similarly, @Santhosh-Sellavel can you please apply to the job posting?

**Comment by Santhosh-Sellavel** [comment]
> **BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:
> 
> * [x]  [@Santhosh-Sellavel] The PR that introduced the bug has been identified. Link to the PR: https://github.com/Expensify/App/pull/15731
> * [x]  [@Santhosh-Sellavel] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: https://github.com/Expensify/App/pull/15731#issuecomment-1563557295
> * [x]  [@Santhosh-Sellavel] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: N/A
> * [x]  [@Santhosh-Sellavel] Determine if we should create a regression test for this bug. Yes just to ensure
> * [c]  [@Santhosh-Sellavel] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
> * [ ]  [@tjferriss] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by Santhosh-Sellavel** [comment]
Regression Test Steps:

1. Open New Dot
2. Logout if already logged in
3. Enter your phone number or email address and click Continue
4. Scroll to the bottom and click Create A New Account or Log In
5. Verify that the keyboard doesn't flicker and Login Page is open again
6. Repeat Step 3
7. Scroll until you can hide the input (Web and desktop can be skipped)
8. Click Go back
9. Verify the page will automatically scroll to the top, and without covering the input

👍  or  👎

**Comment by hellohublot** [comment]
@tjferriss  Similar to [this one](https://github.com/Expensify/App/issues/17200#issuecomment-1530796756), The client always can't see my proposal on upwork, so you can send me the offer directly, https://www.upwork.com/freelancers/~01509ca2c74bfc7073  Thanks.

**Comment by tjferriss** [comment]
Oh duh, I just had to click "Load More". Offers have been sent to both of you. Once they're accepted, I'll process the payment.

**Comment by tjferriss** [comment]
@hellohublot @Santhosh-Sellavel payments have been made.
